stl/inc/deque

```diff
@@ -1236,52 +1322,78 @@ private:
-    iterator insert(const_iterator _Where, _Iter _First, _Iter _Last) {
-        // insert [_First, _Last) at _Where, input iterators
-        size_type _Off = static_cast<size_type>(_Where - begin());
+    template <bool _Bidi, class _It, class _Se>
+    iterator _Insert_range(const size_type _Off, _It _First, _Se _Last) {
+        // insert [_First, _Last) at begin() + _Off
 
 #if _ITERATOR_DEBUG_LEVEL == 2
         _STL_VERIFY(_Mysize() >= _Off, "deque insert iterator outside range");
 #endif // _ITERATOR_DEBUG_LEVEL == 2
 
-        _Adl_verify_range(_First, _Last);
-        auto _UFirst      = _Get_unwrapped(_First);
-        const auto _ULast = _Get_unwrapped(_Last);
+        const size_type _Oldsize = _Mysize();
 
-        size_type _Oldsize = _Mysize();
+        if (_First == _Last) {
+            return begin() + static_cast<difference_type>(_Off);
+        }
 
-        if (_UFirst != _ULast) {
-            if (_Off <= _Mysize() / 2) { // closer to front, push to front then rotate
-                _Restore_old_size_guard<_Pop_direction::_Front> _Guard{this, _Oldsize};
-                for (; _UFirst != _ULast; ++_UFirst) {
-                    emplace_front(*_UFirst); // prepend flipped
+        _Orphan_all();
+        if (_Off <= _Mysize() / 2) { // closer to front, push to front then rotate
+            _Restore_old_size_guard<_Pop_direction::_Front> _Guard{this, _Oldsize};
+            if constexpr (_Bidi) {
+                while (_First != _Last) {
+                    _Emplace_front_internal(*--_Last); // prepend in order
                 }
-
-                _Guard._Container = nullptr;
-
-                size_type _Num = _Mysize() - _Oldsize;
-                _STD reverse(begin(), begin() + static_cast<difference_type>(_Num)); // flip new stuff in place
-                _STD rotate(begin(), begin() + static_cast<difference_type>(_Num),
-                    begin() + static_cast<difference_type>(_Num + _Off));
-            } else { // closer to back
-                _Orphan_all();
-                _Restore_old_size_guard<_Pop_direction::_Back> _Guard{this, _Oldsize};
-                for (; _UFirst != _ULast; ++_UFirst) {
-                    _Emplace_back_internal(*_UFirst);
+            } else {
+                for (; _First != _Last; ++_First) {
+                    _Emplace_front_internal(*_First); // prepend flipped
                 }
+            }
 
-                _Guard._Container = nullptr;
+            _Guard._Container = nullptr;
 
-                _STD rotate(begin() + static_cast<difference_type>(_Off),
-                    begin() + static_cast<difference_type>(_Oldsize), end());
+            const auto _Num = static_cast<difference_type>(_Mysize() - _Oldsize);
+            if constexpr (!_Bidi) {
+                _STD reverse(begin(), begin() + _Num); // flip new stuff in place
             }
+            _STD rotate(begin(), begin() + _Num, begin() + _Num + static_cast<difference_type>(_Off));

```

Comment:
I am slightly unsure about evaluation order here. Should we do
```suggestion
            _STD rotate(begin(), begin() + _Num, begin() + (_Num + static_cast<difference_type>(_Off)));
```

Reply:
Probably, yes, but I'm going to factor out `begin() + _Num` into a temporary instead.